asdf/config.py

```diff
@@ -1,81 +1,237 @@
+    one or more containing `AsdfConfigContext`.
+
+    Returns
+    -------
+    AsdfConfig
+    """
+    if len(_local.config_stack) == 0:
+        return _global_config
+    else:
+        return _local.config_stack[-1]
 
 
 def configure(
     converter_providers=NotSet,
+    validate_on_read=NotSet,
 ):
-    global _config
-    original_config = _config
+    """
+    Update the global asdf configuration.  This method is
+    thread-safe in that it updates the config atomically,
+    but calling this method after threads have been started
+    can lead to race conditions.  Changes affect currently
+    open `AsdfFile` instances.
+
+    The default parameter value `NotSet` leaves that parameter
+    set to its current value.
+
+    Parameters
+    ----------
+    converter_providers : list of asdf.AsdfConverterProvider or None or NotSet
+        `asdf.AsdfConverterProvider` instances to use when reading
+        and writing ASDF files.  If `None`, all providers registered
+        with entry points are used.
+
+    validate_on_read : bool or NotSet
+        If True, ASDF files will be validated against schemas on read.
+    """
+    global _global_config
+
+    _global_config = _global_config._merge(
+        converter_providers=converter_providers,
+        validate_on_read=validate_on_read,
+    )
 
-    if converter_providers is NotSet:
-        converter_providers = _config._converter_providers
-        converters = _config._converters
-        converter_index = _config._converter_index
-    else:
-        converters = None
-        converter_index = None
 
-    _config = AsdfConfig(
+class AsdfConfigContext:
+    """
+    Context manager that temporarily overrides asdf configuration.
+    Users are not intended to construct this object directly; instead,
+    use the `config_context` module method.
+    """
+    def __init__(self, **config_options):
+        self._config_options = config_options
+
+    def __enter__(self):
+        if len(_local.config_stack) == 0:
+            base_config = _global_config
+        else:
+            base_config = _local.config_stack[-1]
+
+        config = base_config._merge(**self._config_options)
+        _local.config_stack.append(config)
+        return config
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        _local.config_stack.pop()
+
+    def __repr__(self):
+        arg_string = ", ".join(["{}={!r}".format(k, v) for k, v in self._config_options if v is not NotSet])
+        return "{}({})".format(self.__class__.__name__, arg_string)
+
+
+def config_context(
+    converter_providers=NotSet,
+    validate_on_read=NotSet,
+):
+    """
+    Create a context manager that temporarily overrides asdf
+    configuration.  Overrides are only applied in the current
+    thread.  Nested calls to this method are applied in order,
+    with the innermost context overriding the others. Changes
+    affect currently open `AsdfFile` instances.
```

Comment:
Hmmm. Is this accurate? I'm wondering if that the configuration may be bound to an open file and persist until it is closed, though that does complicate things. For example, inside a function, I open a file that has a configuration parameter for the inline array threshold, and I set a configuration for a certain threshold before I open it, exit the configuration after opening the file, and return the file. I may have wanted to have that threshold apply so long as the file is open even though the context that the configuration was created has since exited. So I wonder in the open a reference to the configuration can be associated with the AsdfFile instance. Is this crazy thinking? I raise this since I think I was naive in making the analogy to numpy (or numarray where it was conceived) for handling floating point errors. The difference there is the errors arise out of operations performed on arrays, and those operations are bound to the code that is performing them. In this case, I tend to think the behaviors are more bound to the object than the code that is manipulating them. Would we want to change the configuration that applies to a given file at different places in the code?

I still think the current context for configuration still is useful, but I wonder if it should only be associated with where it is executed and not only with the objects created. Sorry for the monkey wrench. They are greasy and slip out my hand often.

Reply:
The behavior could be that the configuration currently set means that is what will be bound to a file if opened or created while in effect, but it can persist with the object.

Reply:
We can do it either way without much trouble I think -- the AsdfFile should be able to grab whatever the current config object is and keep a reference to it.  The individual configs are immutable so that is all it needs to retain the options that were set when the AsdfFile was created.

I've been thinking about this in terms of familiar desktop applications like text editors.  Some settings are editor-level, like disabling spell check, automatically trimming whitespace, etc.  Others are file-level like the file mode (TextEdit has "plain text" and "rich text"), margin width, etc.  Some are both -- you can configure the font size on the file level, but you can also configure the default font size on the editor level.

It seems like we also have some of each category.  We might break ours down like this:

"editor level"
- flag to warn on unrecognized tag
- flag to validate on read
- list of enabled plugins (extensions and the new providers)
- default for memory mapping enabled/disabled
- default ASDF Standard version for new files
- default inline array threshold for new files

"file level"
- ASDF Standard version (overrides default set at editor level)
- memory mapping enabled/disabled (overrides default set at editor level)
- inline array threshold (overrides default set at editor level)

What do you think about breaking down the settings in this way, and keeping the "editor level" settings in asdf.config?  If users changed them in the middle of a session, they'd affect existing AsdfFile instances.  The "file level" settings, by contrast, would be sticky for existing AsdfFile instances.

Reply:
BTW, inline array threshold seems borderline to me -- it's a sort of stylistic choice we make when writing the file, but not really a property of the file itself.  We don't write the `auto_inline` value into the file so other users won't know what threshold was used.

Come to think of it, that's a decent rule of thumb for deciding whether a setting should be "file level" or "editor level": if it gets saved as a property of the actual ASDF file, then it's probably file level, otherwise not.  Though memory mapping enabled/disabled is an obvious exception.

Reply:
The two contexts (editor/file) seem reasonable to me. 

As for inline threshold, that's a bit more murky. I tend to think it is more associated with the file (e.g., someone may want everything inline, or everything smaller than x inline, and they don't want to deal with every level in the code that may add arrays). On the other hand, if they control when the file is written to disk, they could force it there, but will they always control it at that point? I can see variations of where these details are or aren't available to the user (the user may be a developer using someone else's library that adds to the tree, and possibly writes it out. Perhaps we need to think a bit more.

Reply:
As example use cases,

1) someone want to run Roman pipeline (aqueduct?) how would we best allow the user to indicate a inline threshold? In this case they specify an input ASDF file, which is not the output. In this case it simply seems like it should be a option to the pipeline, and thus a global parameter rather than file specific, if running from the command line. If running from a python session they could make it their configuration, and avoid having to specify it every time they execute a pipeline task.

2) They are building their own ASDF product in their code, so they create the file object and manage the tree construction. Either approach is fine.

3) The create or read in a file object then pass it to code that will modify or add to it and writes it out. This seems a bit unusual case but possible. In this event can they be sure that the underlying code won't change the configuration. No, but they don't have any control over that, but they can change the global configuration and hope the code won't override the threshold aspect. If the configuration was attached to the file, I guess the presumption would be that library code should not override it unless there is a strong reason to; so that would argue for allowing it to be attached to the file.

3) another solution may be to provide utilities to copy files to the flavor the user wants and avoid the configuration mess. Inefficient in requiring copies sometime, but straightforward.

I'm still giving this a little more thought.

Reply:
I guess my inclination is to leave it out of file level until we see a strong use case to change that. I don't see one yet.